website/utils/media/processors.py

```diff
@@ -0,0 +1,66 @@
+from da_vinci import Image as DaVinciImage
+from PIL import Image, ImageOps
+
+
+def process_upload(image: DaVinciImage, **kwargs):
+    """Process an incoming image.
+
+    - Rotate and flip the image based on EXIF data.
+    - Resize the image to the maximum allowed size if it is larger.
+    - Convert the image to RGB colors, or keep RGBA for PNGs.
+
+    Warning: for django-thumbnails to save the image with the right filename, the
+    format must also be set on the '<size>.FORMAT' key in settings.THUMBNAILS_SIZES.
+    """
+    # Get the PIL image out of the DaVinci image wrapper,
+    # so we can manipulate it more completely.
+    pil_image = image.get_pil_image()
+
+    # Rotate and flip the image based on EXIF
+    # data, so the image is stored upright.
+    pil_image = ImageOps.exif_transpose(pil_image)
+
+    # Resize the image to at most the maximum allowed size.
+    pil_image.thumbnail(kwargs["size"])
+
+    if kwargs["format"] == "jpg":
+        # Store in RGB colors.
+        pil_image = pil_image.convert("RGB")
+        image.format = "JPEG"
+    elif kwargs["format"] == "png":
+        image.format = "PNG"
+        if pil_image.mode not in ("RGBA", "RGB"):
+            pil_image = pil_image.convert("RGB")
+    else:
+        raise ValueError(f"Unsupported format: {kwargs['format']}")
+
+    # Put the PIL image back into the wrapper.
+    image.set_pil_image(pil_image)
+    return image
+
+
+def thumbnail(image: DaVinciImage, **kwargs):
+    """Thumbnail an image to at most the given size.
+
+    If `kwargs["cover"]`, the image is cropped to the aspect ratio of the given size.
+    The thumbnails are saved to WebP format with lossy compression.
+
+    Warning: for django-thumbnails to save the image with the right filename,
+    the '<size>.FORMAT' key in settings.THUMBNAILS_SIZES must be set to 'webp'.
+
+    """
+    pil_image = image.get_pil_image()
+
+    size = kwargs["size"]
+    if not kwargs["cover"]:
+        ratio = min(a / b for a, b in zip(size, pil_image.size))
+        size = tuple(int(ratio * x) for x in pil_image.size)
```

Comment:
I'm not sure, but I think I found why some thumbnails are really low quality, and this has the same problem.

The problem is as follows: some users have very non-square profile pictures. (e.g. 16:9 and I even found one that was more than 2:1 portrait). The thumbnailing code resizes it then to be e.g. 300px along the largest axis. This results in images of e.g. 300*200px or 145*300px.

When the time to display the thumbnail comes, the website uses square card views. This means that images get cropped by the browser. But because the thumbnail gets cropped, we get effective resolutions of 200*200px or even 145*145px.

We should probably fix this by cropping the thumbnails beforehand, either as a separate type of thumbnail (because I can imagine we also have use-cases for non-cropped thumbnails, although I'm not sure), or making the cropping along the shortest axis (which might not be a great idea, as images can then be larger than expected).

Reply:
That's the case for sizes without cover=True. But that has actually also been the case before djangp-thumbnails (I coppied this size calculation from our old custom thumbnailer). I did notice, in the django-thumbnails settings from before this, that the sizes had some weird stretching settings.

So with this PR, the 'small' thumbnails are cropped to cover 300x300 (only less if the original smallest dimension was <300px). Only the `photo_medium`, `photo_large` and `source` sizes are resized to _fit_ in some size (and hence potentially much smaller)

Reply:
This was indeed the cause for [#2936](https://github.com/svthalia/concrexit/issues/2936). One of the examples in the screenshot there is indeed 199x300, instead of 300x300. But with this PR small does get cropped.